lib/resty/waf/collections.lua

```diff
@@ -70,45 +70,52 @@ _M.lookup = {
 		collections.STATUS           = ngx.status
 	end,
 	body_filter = function(waf, collections, ctx)
-		if ctx.buffers == nil then
-			ctx.buffers  = {}
-			ctx.nbuffers = 0
-		end
-
-		local data  = ngx.arg[1]
-		local eof   = ngx.arg[2]
-		local index = ctx.nbuffers + 1
-
-		local res_length = tonumber(collections.RESPONSE_HEADERS["content-length"])
-		local res_type   = collections.RESPONSE_HEADERS["content-type"]
-
-		if not res_length or res_length > waf._res_body_max_size then
-			ctx.short_circuit = not eof
-			return
-		end
-
-		if not res_type or not util.table_has_key(res_type, waf._res_body_mime_types) then
-			ctx.short_circuit = not eof
-			return
-		end
-
-		if data then
-			ctx.buffers[index] = data
-			ctx.nbuffers = index
-		end
-
-		if not eof then
-			-- Send nothing to the client yet.
-			ngx.arg[1] = nil
-
-			-- no need to process further at this point
-			ctx.short_circuit = true
-			return
-		else
-			collections.RESPONSE_BODY = table_concat(ctx.buffers, '')
-			ngx.arg[1] = collections.RESPONSE_BODY
-		end
-	end,
+                if ctx.buffers == nil then
+                        ctx.buffers  = {}
+                        ctx.nbuffers = 0
+                        ctx.res_length = 0
+                        ctx.skip_buffering = false
+                end
+
+                local data  = ngx.arg[1]
+                local eof   = ngx.arg[2]
+                local index = ctx.nbuffers + 1
+
+                local res_type   = collections.RESPONSE_HEADERS["content-type"]
+
+                if not res_type or not util.table_has_key(res_type, waf._res_body_mime_types) then
+                        ctx.short_circuit = not eof
+                        return
+                end
+
+                if ctx.skip_buffering then
+                        ctx.short_circuit = true
+                        return
+                else
+                        ngx.arg[1] = nil
+                end
+
+                if data then
+                        ctx.buffers[index] = data
+                        ctx.nbuffers = index
+                        ctx.res_length = ctx.res_length + string.len(data)
+                end
+
+                if eof or ctx.res_length > waf._res_body_max_size then
```

Comment:
when `ctx.res_length > waf._res_body_max_size` is true the response will be the concatenation of all chunks buffered.
The next chunk will not be processed because of `ctx.skip_buffering` on line 91: this means that ngx.arg[1] won't be touched, ctx.short_circuit=true will be applied and chunk will be sent to client without lost of datas

Reply:
Thanks. Do you have a test case to demonstrate this behavior?

We will need regression and integration tests for this change anyhow. Thanks!

Reply:
Any progress with this?

Reply:
@sandromodarelli im still waiting on a response for the following:

> Do you have a test case to demonstrate this behavior?
> 
> We will need regression and integration tests for this change anyhow. Thanks!